File: java/src/main/java/io/github/gsoultan/anubis/Permissions.java
package io.github.gsoultan.anubis;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

/**
 * An effective permission set — what a caller may do, expanded from every role
 * they hold.
 */
public final class Permissions implements Iterable<Permission> {
    private static final Permissions EMPTY = new Permissions(List.of());

    private final List<Permission> items;

    private Permissions(List<Permission> items) {
        this.items = List.copyOf(items);
    }

    public static Permissions empty() {
        return EMPTY;
    }

    public static Permissions of(List<String> permissions) {
        if (permissions == null || permissions.isEmpty()) {
            return EMPTY;
        }
        List<Permission> out = new ArrayList<>(permissions.size());
        for (String p : permissions) {
            out.add(Permission.of(p));
        }
        return new Permissions(out);
    }

    public boolean has(Permission permission) {
        return items.contains(permission);
    }

    public boolean has(String permission) {
        return has(Permission.of(permission));
    }

    public boolean hasAny(String... permissions) {
        for (String p : permissions) {
            if (has(p)) {
                return true;
            }
        }
        return false;
    }

    public Permissions ofApp(String app) {
        List<Permission> out = new ArrayList<>();
        for (Permission p : items) {
            if (p.app().equals(app)) {
                out.add(p);
            }
        }
        return new Permissions(out);
    }

    public List<String> toStrings() {
        List<String> out = new ArrayList<>(items.size());
        for (Permission p : items) {
            out.add(p.key());
        }
        return Collections.unmodifiableList(out);
    }

    public int size() {
        return items.size();
    }

    @Override
    public Iterator<Permission> iterator() {
        return items.iterator();
    }
}

File: php/src/Permissions.php
<?php

declare(strict_types=1);

namespace Anubis;

/**
 * An effective permission set — what a caller may do, expanded from every role
 * they hold.
 *
 * @implements \IteratorAggregate<int, Permission>
 */
final readonly class Permissions implements \IteratorAggregate, \Countable
{
    /** @var list<Permission> */
    private array $items;

    /** @param iterable<Permission|string> $permissions */
    public function __construct(iterable $permissions = [])
    {
        $out = [];
        foreach ($permissions as $p) {
            $out[] = Permission::of($p);
        }
        $this->items = $out;
    }

    public function has(Permission|string $permission): bool
    {
        $want = Permission::of($permission);
        foreach ($this->items as $p) {
            if ($p->equals($want)) {
                return true;
            }
        }
        return false;
    }

    public function hasAny(Permission|string ...$permissions): bool
    {
        foreach ($permissions as $p) {
            if ($this->has($p)) {
                return true;
            }
        }
        return false;
    }

    /** Narrow to one application's permissions. */
    public function ofApp(string $app): self
    {
        return new self(array_filter($this->items, static fn (Permission $p): bool => $p->app() === $app));
    }

    /** @return list<string> */
    public function toStrings(): array
    {
        return array_map(static fn (Permission $p): string => $p->key, $this->items);
    }

    public function getIterator(): \Traversable
    {
        return new \ArrayIterator($this->items);
    }

    public function count(): int
    {
        return count($this->items);
    }
}
